File: LeetCode/remote/symmetric_tree.cs
//  https://leetcode.com/problems/symmetric-tree/
//
//  Given a binary tree, check whether it is a mirror of itself (ie, symmetric around its center).
//
//  For example, this binary tree is symmetric:
//

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public bool IsSymmetric(TreeNode root) {
        if (root == null)
        {
            return true;
        }
        
        return IsSymmetric(root.left, root.right);
    }
    
    public bool IsSymmetric(TreeNode first, TreeNode second)
    {
        if (first == null || second == null)
        {
            return first == null && second == null;
        }
        
        return first.val == second.val && IsSymmetric(first.left, second.right) && IsSymmetric(first.right, second.left);
    }
}

File: LeetCode/remote/symmetric_tree.rb
#   https://leetcode.com/problems/symmetric-tree/
#   Given a binary tree, check whether it is a mirror of itself (ie, symmetric around its center).
#   https://leetcode.com/submissions/detail/55551974/
#
#   Submission Details
#   192 / 192 test cases passed.
#       Status: Accepted
#       Runtime: 100 ms
#           
#           Submitted: 0 minutes ago
def is_symmetric(root)
    root.nil? ? true : symmetric(root.left, root.right)
end

def symmetric(first, second)
    return first.nil? && second.nil? if first.nil? || second.nil? 
    return first.val == second.val &&
           symmetric(first.left, second.right) &&
           symmetric(first.right, second.left)
end
